adr/ADR-42.md

```diff
@@ -0,0 +1,165 @@
+behaviours configured using the `PriorityPolicy`.
+
+Technically in some of the proposed policies the `PriorityGroups` have no real meaning today, but we keep it for consistency 
+and allow for future features to be added that would be per-group without then requiring all clients to be updated. 
+Future message grouping features would require groups to be listed here.
+
+In the initial implementation we should limit `PriorityGroups` to one per consumer only and error should one be made with 
+multiple groups. In future iterations multiple groups will be supported along with dynamic partitioning of stream
+data.
+
+This is only supported on Pull Consumers, configuring this on a Push consumer must raise an error.
+
+## Priority Policies
+
+### `overflow` policy
+
+Users want certain clients to only get messages when certain criteria are met.
+
+Imagine jobs are best processed locally in `us-east-1` but at times there might be so many jobs in that region that 
+`us-west-1` region handling some overflow, while less optimal in terms of transit costs and latency, would be desirable 
+to ensure serving client needs as soon as possible.
+
+The `overflow` policy enables Pull requests to be served only if criteria like `num_pending` and `num_ack_pending` are
+above a certain limit, otherwise those Pull requests will sit idle in the same way that they would if no messages were
+available (receiving heartbeats etc).
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "overflow",
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+Here we state that the Consumer has one group called `jobs`, it is operating on the `overflow` policy and requires `explicit`
+Acks, any other Ack policy will produce an error. If we force this ack policy in normal use we should error in Pedantic mode.
+
+Pull requests will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"min_pending": 1000` - only deliver messages when `num_pending` for the consumer is >= 1000
+ * `"min_ack_pending: 1000` - only deliver messages when `ack_pending` for the consumer is >= 1000
+
+If `min_pending` and `min_ack_pending` are both given either being satisfied will result in delivery (boolean OR).
+
+In the specific case where MaxAckPending is 1 and a pull is made using `min_pending: 1` this should only be served when
+there are no other pulls waiting. This means we have to give priority to pulls without conditions over those with when
+considering the next pull that will receive a message.
+
+### `pinned_client` policy
+
+Users want to have a single client perform all the work for a consumer, but they also want to have a stand-by client that
+can take over when the primary, aka `pinned` client, fails.
+
+**NOTE: We should not describe this in terms of exclusivity as there is no such guarantee, there will be times when one
```

Comment:
I use double acks in my example for consumer groups for that very reason (i.e. if the double ack returns an error then 'rollback' rather than 'commit')

Reply:
@jnmoyne how do you think double ack help you here?
Double acking will not even return error if you ack outside of the ack_wait window, or done from different puller.

AFAIK server never sends back an error in double ack. It's the client timing out/getting no responders being the only cases for errors.

There is a related issue here: https://github.com/nats-io/nats-server/issues/5113

Reply:
This is the failure scenario I could reproduce with my consumer group code and example: the active fetcher gets a message (in my example max acks pending is 1, as it's kind of the whole point with consumer groups :)) and gets suspended while processing the message (similar to if it gets isolated because for example of a network split-brain), while it is suspended another client becomes the active fetcher and start receiving (and processing) messages, starting with the one message that had been delivered once to the previous (now suspended) active client. So the new guy gets that message and processes it and (double) acks it and it all goes well, and it goes on to receive and process further messages. Then the previous active client gets un-suspended continues processing its message and then acks it (though it was actually processed and acked by the new active client in the mean time), which is a problem. 

So what I show in my example is how (if your processing at the end involves writing the result of the processing to a transaction-capable system like an SQL DB for example) you can avoid that problem and get 'exactly once processing'. What the client code does is: upon receiving the message is starts the transaction, then does it processing and prepares the transaction. At that point it does a double-ack on the message, if the message had been acked before by someone else (i.e. the client that became active while this one was suspended or isolated) then the double-ack fails and it does a rollback, if the double ack returns successfully then it does a commit instead.

Reply:
Server **will not** send you back error if you ack the same message second time.

Client might check for double acking the same message twice, but that's it. It would not work across clients/connections.

Hence you cannot rely on that mechanism.

In fact, server **never** responds with an error, unless it's `no responders`, which happens only if jetstream is disabled.

You can even mess up stream sequences, delivert sequences etc, and server will happily respond with ok.

Reply:
In my experimentation the double ack would return an error when I caused another client to become active (and therefore get re-delivery of the message and acking it) and it gets un-suspended and tries to double ack the same message.

Never realized that the server did not return an error if the message is already acked, and it certainly fooled me into thinking that the server did that.

In any case I would consider it a bug that the server doesn't return an error to an ack request if the message in question has already been acked. To me that's a bit the whole point of a double ack (besides knowing that the ack message has made it to the consumer's leader): knowing that the ack was indeed successful or not, and I would consider trying to ack an already ack'd message as an error (I would even go as far as expecting an error when trying to ack with a delivery count that doesn't match the current delivery count for the message).

Reply:
Would be interested in knowing what error you get, anyone can ack any messages - delivered to that client or not, already acked or not, ack wait expired or not etc.

It's an area of big concern for exactly once behaviors imo

Reply:
@ripienaar Go client tracks if given message was already acked, and will error if you ack it second time. Obviously works only for acks done by the same client instance. To be fair - I dont like that behavior considering how little info you actually get from the server and that it might be necessary to ack it again in some failure scenarios that server did not inform about.

Reply:
Oh yes that's definitely not behaviour we need in the client imo, it will be confusing as we saw but also leed to a false sense of whats possible when things only work within one client connection.

Reply:
I checked and the actual error I get is a 'no responders available' - probably more an effect of the way I trigger the failure scenario. 

In any case I agree (I think we all agree) that this current behavior of not returning an error if already acked (or allowing to ack a message not even delivered) is not right and does have implications for 'exactly once' behaviors.